plutus-context-builder/src/Test/Plutus/ContextBuilder/Internal.hs

```diff
@@ -288,79 +294,131 @@ data Minting
       Show
     )
 
-{- | A way to incrementally build up a script context.
+{- | Indicates whether a 'ContextBuilder' has named components or not.
 
- It is tagged with a 'Purpose' as a marker for what kind of 'ScriptPurpose'
- will be used in the constructed 'ScriptContext'.
+ @since 2.0
+-}
+data Naming = Anonymous | Named
```

Comment:
I am _strongly_ opposes to type synonyms, not least of all for reasons of error reporting. If someone wants to create one, they can do so themselves. Abbreviation is possible, but I'd rather the clarity of a slightly longer name.

Reply:
I don't think adding the distinction between `Anonymous` and `Named` context builders at the type level will be a lot helpful; specifically we can't union anonymous builders, or union an anonymous builder with a named one. I think a structure that stores both one anonymous fragment *and* a map of named fragments will be simpler to use.

Reply:
One of the main reasons I went for this distinction is to avoid 'double naming' - you can name an anonymous builder, but then it's named. Furthermore, the behaviour of a named and an anonymous builder under `<>` is different, and thus, keeping them type-level separate ensures you don't run face-first into this, as @blamario did in the original issue.

Reply:
I see. It's a bit confusing to make the `(<>)` behavior different for named and unnamed variants though, but I guess that's to make breakage minimal.

Reply:
The entire initial issue stemmed from unexpected behaviour of `<>`, which is why this took the form that it did. I'm definitely not 100% happy with this solution either, so if there's some way to do this better, I'd love suggestions.

Reply:
What I [originally imagined](https://github.com/Liqwid-Labs/plutus-extra/issues/193#issuecomment-1062285274) was to have the anonymous builder equivalent to named builders with an empty name, i.e. `named "" == id`. Naming would then simply add a prefix to all the keys. The behaviour of `<>` would be uniform, but still potentially surprising - not to me, but I may have a weird mind.

Reply:
Like @blamario, I think it is better to have consistent `(<>)` behavior on both anonymous and named variants - both concats the `Seq`s for each component. This is definitely less confusing, and we can rename the current `unionNamed` to `unionNamedWith` and let `unionNamed = unionNamedWith const` (i.e. the current behavior of `(<>)`).

Reply:
I didn't realize that, thanks for pointing it out. I added a comment to the code.